File: HOCMIM.py
import numpy as np
from sklearn.metrics import mutual_info_score


def mutual_info(x, y):
    """
    Compute mutual information between two variables.

    Parameters:
    x (np.ndarray): Feature vector.
    y (np.ndarray): Target vector.

    Returns:
    float: Mutual information.
    """
    return mutual_info_score(x, y)


def conditional_mutual_info(x, y, z):
    """
    Compute conditional mutual information I(x; y | z).

    Parameters:
    x (np.ndarray): Feature vector.
    y (np.ndarray): Target vector.
    z (np.ndarray): Conditioning variable.

    Returns:
    float: Conditional mutual information.
    """
    xyz = np.column_stack((x, z))
    yz = np.column_stack((y, z))

    mi_xyz = mutual_info(xyz.ravel(), y)
    mi_z = mutual_info(z.ravel(), y)

    return mi_xyz - mi_z


def n_order_total_redundancy_max(xk, s, y, order):
    """
    Compute n-order total redundancy maximum.

    Parameters:
    xk (np.ndarray): Candidate feature.
    s (np.ndarray): Selected features.
    y (np.ndarray): Target vector.
    order (int): Maximum redundancy order.

    Returns:
    tuple: Maximum redundancy and iteration order.
    """
    ns = s.shape[1] if s.ndim > 1 else 0
    max_rn = -np.inf

    selected_indices = []
    rn_max = mutual_info(xk, y)

    for i in range(min(ns, order)):
        max_rn = -np.inf
        for j in range(ns):
            if j not in selected_indices:
                mi = mutual_info(xk, np.hstack([s[:, selected_indices], s[:, j:j+1]]))
                cmi = conditional_mutual_info(xk, s[:, j], y)
                if mi - cmi > max_rn:
                    max_rn = mi - cmi
                    best_index = j

        selected_indices.append(best_index)
        if max_rn / rn_max < 0.05:
            break

    return max_rn, len(selected_indices)


def hocmim(x, y, k, n, verbose=False):
    """
    High-order Conditional Mutual Information Maximization (HOCMIM) for feature selection.

    Parameters:
    x (np.ndarray): Training data (samples x features).
    y (np.ndarray): Target labels.
    k (int): Number of features to select.
    n (int): Order of HOCMIM approximation.
    verbose (bool): Print progress if True.

    Returns:
    tuple: Selected features, elapsed time, CMI approximations, and selection order.
    """
    import time

    start_time = time.time()
    n_features = x.shape[1]
    k = min(k, n_features)

    # Mutual information for all features
    mi = np.array([mutual_info(x[:, i], y) for i in range(n_features)])

    # Initialize selected features
    selected = []
    cmi_app = []
    order = []

    for t in range(k):
        max_cmi = -np.inf
        best_feature = None

        for i in range(n_features):
            if i not in selected:
                redundancy, it_order = n_order_total_redundancy_max(
                    x[:, i], x[:, selected], y, n
                )
                cmi = mi[i] - redundancy
                if cmi > max_cmi:
                    max_cmi = cmi
                    best_feature = i

        if best_feature is not None:
            selected.append(best_feature)
            cmi_app.append(max_cmi)
            order.append(it_order)

        if verbose:
            print(f"Step {t+1}/{k}: Selected feature {best_feature}")

    elapsed_time = time.time() - start_time
    return selected, elapsed_time, cmi_app, order

File: HOCMIM.m
function [S, tElapsed, CMIapp, par, Order] = HOCMIM(X, Y, K, n, fast, verbose)
    % Parallel implementation of HOCMIM feature selection
    % Author: Francisco Alexandre
    % Date: 11/2020, Version: 1.0

    % Check input arguments
    if nargin < 6
        verbose = false;
    end

    % Initialize variables
    N = size(X, 2);
    if K > N
        K = N;
    end
    tStart = tic; % Start timer
    A = 1:N; % Feature indices
    S = zeros(1, K); % Selected features
    CMIapp = zeros(1, K); % Conditional mutual information approximation

    % Compute mutual information for the first feature
    MI = arrayfun(@(i) mutual_info(X(:, i), Y), 1:N);
    [max_MI, max_idx] = max(MI);

    % Initialize selected features and order
    S(1) = A(max_idx);
    CMIapp(1) = max_MI;
    Order = zeros(1, K);
    Order(1) = 1;
    A(max_idx) = []; % Remove selected feature

    if verbose
        fprintf('Selected features: #%d\n', S(1));
    end

    % Iteratively select features
    for t = 2:K
        CMI = zeros(1, length(A));
        ord = zeros(1, length(A));

        parfor i = 1:length(A)
            [redundancy, ord(i)] = n_order_total_redundancy_max(X(:, A(i)), X(:, S(1:t-1)), Y, n);
            CMI(i) = MI(A(i)) - redundancy;
        end

        [~, rank_idx] = max(CMI);
        S(t) = A(rank_idx);
        Order(t) = ord(rank_idx);
        CMIapp(t) = CMI(rank_idx);
        A(rank_idx) = []; % Remove selected feature

        if verbose
            fprintf('Selected features: %s\n', mat2str(S(1:t)));
        end
    end

    tElapsed = toc(tStart); % Compute elapsed time
    par.S = S;
    par.tElapsed = tElapsed;
    par.CMIapp = CMIapp;
end

function [redundancy, it_order] = n_order_total_redundancy_max(Xk, S, Y, order)
    % Compute n-order total redundancy maximum
    ns = size(S, 2);
    max_Rn = -Inf;
    ind_selected = [];

    % First order
    for i = 1:ns
        mi_ = mutual_info(Xk, S(:, i));
        cmi_ = cond_mutual_info(Xk, S(:, i), Y);
        if mi_ - cmi_ > max_Rn
            max_Rn = mi_ - cmi_;
            best_idx = i;
        end
    end
    ind_selected = [ind_selected, best_idx];
    redundancy = max_Rn;
    it_order = 1;

    % Higher orders
    for j = 2:min(ns, order)
        max_Rn = -Inf;
        for i = 1:ns
            if ~ismember(i, ind_selected)
                mi_ = mutual_info(Xk, [S(:, i), S(:, ind_selected)]);
                cmi_ = cond_mutual_info(Xk, [S(:, i), S(:, ind_selected)], Y);
                if mi_ - cmi_ > max_Rn
                    max_Rn = mi_ - cmi_;
                    best_idx = i;
                end
            end
        end
        ind_selected = [ind_selected, best_idx];
        redundancy = max_Rn;
        if redundancy / max_Rn < 0.05
            break;
        end
    end
end

function MI = mutual_info(X, Y)
    % Compute mutual information between X and Y
    if size(X, 2) > 1
        X = MIToolboxMex(3, X);
    end
    if size(Y, 2) > 1
        Y = MIToolboxMex(3, Y);
    end
    [p12, p1, p2] = estpab(X, Y);
    MI = estmutualinfo_indjs(p12, p1, p2, size(X, 1));
    if isnan(MI) || MI < 0
        MI = 0;
    end
end

function CMI = cond_mutual_info(X, Y, Z)
    % Compute conditional mutual information
    MI_XZ = mutual_info(X, [Y, Z]);
    MI_YZ = mutual_info(Y, Z);
    CMI = MI_XZ - MI_YZ;
    if isnan(CMI) || CMI < 0
        CMI = 0;
    end
end
